### ISSUE TITLE
Investigate climbing database CPU on hush-house

### ISSUE THREAD

**Issue by clarafu** [concourse/concourse#5577]
## Summary

On the large deployment (called `hush-house`) that the concourse team manages, we have observed the database CPU being pegged at around 95%. It seemed to be a slow climb over a couple of weeks. The image below shows the DB cpu for the deployment, starting from when we first deployed v6.0 until the current date. The small dip that is observed on the graph near the beginning of May is due to a deploy of v6.1 onto `hush-house`. This dip in CPU seems to climb back up to 95% within the next week. 

![image](https://user-images.githubusercontent.com/19290703/81574770-3b622a80-9374-11ea-9cbd-cd3a44aad77f.png)

### Things to look into:
- [x] What heavy queries are being run on the database? (https://github.com/concourse/concourse/wiki/Schema-Inspection-Queries)
- [ ] Take some heap, stack, CPU dumps from the web nodes

## Context

* What version is it? v6.1

**Comment by clarafu** [comment]
Running the schema inspection query for finding the current in-flight queries and also the slowest queries, we saw that the first occurrence query seemed to be a bottleneck for the database.

```
 WITH builds_of_job AS ( SELECT id FROM builds WHERE job_id = $1 ) SELECT EXISTS ( SELECT ? FROM build_resource_config_version_inputs i JOIN builds_of_job b ON b.id = i.build_id WHERE i.name = $2 AND i.version_md5 = $3 )
```

![image](https://user-images.githubusercontent.com/19290703/81580782-df9b9f80-937b-11ea-9190-3e29d0a2ddaa.png)

It was taking 2 seconds for each first occurrence query to finish and had the highest total time because of the number of times it is run. 

Another query we found interesting is the query to gc build image resource caches. It had the highest avg time to run but is being run a lot less than the first occurrence query.

```
DELETE FROM build_image_resource_caches birc USING builds b WHERE birc.build_id = b.id AND build_id < $1 AND b.job_id = $2
```

**Comment by aoldershaw** [comment]
Could you share the schema inspection query you're using to find these stats?

**Comment by clarafu** [comment]
Working on the first occurrence query, we figured out that it was taking approximately 12 seconds for the following query to run on `hush-house` 

```
EXPLAIN ANALYZE WITH builds_of_job AS ( SELECT id FROM builds WHERE job_id = '123389' ) SELECT EXISTS ( SELECT 1 FROM build_resource_config_version_inputs i JOIN builds_of_job b ON b.id = i.build_id WHERE i.name = '10m-time' AND i.version_md5 = '06e85a416753b5c0525b28348793914e');
```

and it produces the following explain analyzer result

```
Result  (cost=178540.84..178540.85 rows=1 width=1) (actual time=2403.589..2403.590 rows=1 loops=1)
  CTE builds_of_job
    ->  Index Only Scan using builds_ordered_by_job_id_idx on builds  (cost=0.43..57243.49 rows=54513 width=4) (actual time=0.032..0.036 rows=2 loops=1)
          Index Cond: (job_id = 123389)
          Heap Fetches: 5
  InitPlan 2 (returns $1)
    ->  Nested Loop  (cost=0.00..121297.36 rows=1 width=0) (actual time=2403.588..2403.588 rows=1 loops=1)
          Join Filter: (i.build_id = b.id)
          Rows Removed by Join Filter: 1
          ->  Seq Scan on build_resource_config_version_inputs i  (cost=0.00..119525.68 rows=1 width=4) (actual time=2403.541..2403.541 rows=1 loops=1)
                Filter: ((name = '10m-time'::text) AND (version_md5 = '06e85a416753b5c0525b28348793914e'::text))
                Rows Removed by Filter: 4516741
          ->  CTE Scan on builds_of_job b  (cost=0.00..1090.26 rows=54513 width=4) (actual time=0.034..0.039 rows=2 loops=1)
Planning time: 0.358 ms
Execution time: 2403.631 ms
```

The slowest part of the query seemed to be the sequence scan on the `build_resource_config_version_inputs` table where it was filtered on `name` and `version_md5`. This was a little surprising because this filtering should have been using an index. Looking at the existing indexes on the table, there was only a unique index on `build_id`, `resource_id`, `version_md5` and name and also another index on `resource_id` and `version_md5`. Since the filtering is only on `name` and `version_md5`, it won't use any of the existing indexes. 

Knowing that, it would be preferable for the first occurrence query to use an existing index rather than adding a new one because each index on the `build_resource_config_version_inputs` table is just over 400 MBs with how large the table is. We can make the query use the index on `resource_id` and `version_md5` by adding a filtering on `resource_id` to the query. 

```
EXPLAIN ANALYZE WITH builds_of_job AS ( SELECT id FROM builds WHERE job_id = '123389' ) SELECT EXISTS ( SELECT 1 FROM build_resource_config_version_inputs i JOIN builds_of_job b ON b.id = i.build_id WHERE i.name = '10m-time' AND i.version_md5 = '06e85a416753b5c0525b28348793914e' AND i.resource_id = '41573');
```

After adding a where clause for the `resource_id`, the query runs in 20 ms.

```
Result  (cost=58577.34..58577.35 rows=1 width=1) (actual time=21.374..21.374 rows=1 loops=1)
  CTE builds_of_job
    ->  Index Only Scan using builds_ordered_by_job_id_idx on builds  (cost=0.43..57273.81 rows=54523 width=4) (actual time=0.072..0.089 rows=3 loops=1)
          Index Cond: (job_id = 123389)
          Heap Fetches: 3
  InitPlan 2 (returns $1)
    ->  Hash Join  (cost=8.59..1303.52 rows=1 width=0) (actual time=21.373..21.373 rows=1 loops=1)
          Hash Cond: (b.id = i.build_id)
          ->  CTE Scan on builds_of_job b  (cost=0.00..1090.46 rows=54523 width=4) (actual time=0.091..0.109 rows=3 loops=1)
          ->  Hash  (cost=8.58..8.58 rows=1 width=4) (actual time=21.252..21.252 rows=1 loops=1)
                Buckets: 1024  Batches: 1  Memory Usage: 9kB
                ->  Index Scan using build_inputs_resource_versions_idx on build_resource_config_version_inputs i  (cost=0.56..8.58 rows=1 width=4) (actual time=21.247..21.248 rows=1 loops=1)
                      Index Cond: ((resource_id = 41573) AND (version_md5 = '06e85a416753b5c0525b28348793914e'::text))
                      Filter: (name = '10m-time'::text)
Planning time: 0.327 ms
Execution time: 21.415 ms
```

**Comment by pivotal-jamie-klassen** [comment]
> `Execution time: 2403.631 ms`
> ...
> `Execution time: 21.415 ms`

😱 🔥 🎉

**Comment by clarafu** [comment]
@taylorsilva and I looked into the query that is currently taking the longest in avg time spent for running it. It seemed to be the query for deleting build image resource caches for any builds older than the currently finished successful build.

```
EXPLAIN ANALYZE DELETE FROM build_image_resource_caches birc USING builds b WHERE birc.build_id = b.id AND build_id < 4128796 AND b.job_id = 123389;
```

If we run this query against a job that has only a few builds, it runs pretty fast. We ran it against a job that had 26 builds and it produced the following result. It is able to run the delete in 0.1 second.

```
Delete on build_image_resource_caches birc  (cost=0.85..3686.06 rows=1 width=12) (actual time=102.985..102.985 rows=0 loops=1)
  ->  Nested Loop  (cost=0.85..3686.06 rows=1 width=12) (actual time=102.914..102.919 rows=1 loops=1)
        ->  Index Scan Backward using builds_ordered_by_job_id_idx on builds b  (cost=0.43..1078.16 rows=318 width=10) (actual time=6.497..71.136 rows=27 loops=1)
              Index Cond: (job_id = 6202)
        ->  Index Scan using build_image_resource_caches_build_id on build_image_resource_caches birc  (cost=0.42..8.19 rows=1 width=10) (actual time=1.175..1.175 rows=0 loops=27)
              Index Cond: ((build_id = b.id) AND (build_id < 362730))
Planning time: 0.395 ms
Execution time: 103.018 ms
```

We then ran the query against a job that had a large number of builds (around 54000 builds) and it got significantly slower, taking 99 seconds to run.

```
Delete on build_image_resource_caches birc  (cost=125555.55..136874.42 rows=9749 width=12) (actual time=99117.848..99117.849 rows=0 loops=1)
  ->  Hash Join  (cost=125555.55..136874.42 rows=9749 width=12) (actual time=99117.846..99117.846 rows=0 loops=1)
        Hash Cond: (birc.build_id = b.id)
        ->  Seq Scan on build_image_resource_caches birc  (cost=0.00..9273.38 rows=519467 width=10) (actual time=0.016..627.923 rows=520378 loops=1)
              Filter: (build_id < 4128796)
              Rows Removed by Filter: 124
        ->  Hash  (cost=124896.90..124896.90 rows=52692 width=10) (actual time=98325.016..98325.016 rows=54143 loops=1)
              Buckets: 65536  Batches: 1  Memory Usage: 2839kB
              ->  Bitmap Heap Scan on builds b  (cost=1252.79..124896.90 rows=52692 width=10) (actual time=858.413..98262.866 rows=54143 loops=1)
                    Recheck Cond: (job_id = 123389)
                    Heap Blocks: exact=39850
                    ->  Bitmap Index Scan on builds_ordered_by_job_id_idx  (cost=0.00..1239.62 rows=52692 width=0) (actual time=784.373..784.373 rows=54145 loops=1)
                          Index Cond: (job_id = 123389)
Planning time: 0.323 ms
Execution time: 99117.996 ms
```

We theorized that as the number of builds that a job has grows, the query will continue to get slower because of the way that it filters on the builds of the job and also the builds that are older than the current finishing build. These filters will mostly result in returning back almost all the builds for the job because we are most likely finishing the latest or one of the later builds, therefore the filter on the builds older than the `build_id` passed to it is probably not filtering out many rows. Chances are that a lot of the resulting builds of the filtering on `job_id` and < `build_id` won't have any caches in the `build_image_resource_caches` table because we are constantly removing them each time a build finishes, therefore the slowness that comes from filtering all the rows is mostly wasted at the end of the query execution. 

In order to make this query faster, we tried to first figure out all the build image resource caches that are associated to a build of the `job_id` and then use those resulting caches to filter down to caches that belong to builds older than the `build_id`.

```
EXPLAIN ANALYZE WITH foo AS (SELECT resource_cache_id, build_id FROM build_image_resource_caches brc JOIN builds b ON b.id = brc.build_id WHERE b.job_id = 123389) 
DELETE FROM build_image_resource_caches birc USING foo f WHERE f.build_id = birc.build_id AND birc.build_id < 4129151;
```

We ran the optimized query against the same job that had 54000 builds and it resulted in the query being run in 1.4 seconds.

```
Delete on build_image_resource_caches birc  (cost=30920.98..35827.09 rows=10308 width=34) (actual time=1396.282..1396.282 rows=0 loops=1)
  CTE foo
    ->  Hash Join  (cost=3215.49..13039.03 rows=9484 width=8) (actual time=822.970..872.704 rows=1 loops=1)
          Hash Cond: (brc.build_id = b.id)
          ->  Seq Scan on build_image_resource_caches brc  (cost=0.00..7833.60 rows=505360 width=8) (actual time=0.037..115.987 rows=520398 loops=1)
          ->  Hash  (cost=2556.84..2556.84 rows=52692 width=4) (actual time=368.229..368.229 rows=54147 loops=1)
                Buckets: 65536  Batches: 1  Memory Usage: 2416kB
                ->  Index Only Scan Backward using builds_ordered_by_job_id_idx on builds b  (cost=0.43..2556.84 rows=52692 width=4) (actual time=0.021..193.304 rows=54147 loops=1)
                      Index Cond: (job_id = 123389)
                      Heap Fetches: 2765
  ->  Hash Join  (cost=17881.95..22788.06 rows=10308 width=34) (actual time=1396.280..1396.280 rows=0 loops=1)
        Hash Cond: (f.build_id = birc.build_id)
        ->  CTE Scan on foo f  (cost=0.00..189.68 rows=9484 width=32) (actual time=822.983..872.722 rows=1 loops=1)
        ->  Hash  (cost=9097.00..9097.00 rows=505356 width=10) (actual time=496.419..496.419 rows=520298 loops=1)
              Buckets: 131072  Batches: 8  Memory Usage: 3825kB
              ->  Seq Scan on build_image_resource_caches birc  (cost=0.00..9097.00 rows=505356 width=10) (actual time=0.025..251.971 rows=520298 loops=1)
                    Filter: (build_id < 4129151)
                    Rows Removed by Filter: 100
Planning time: 0.476 ms
Execution time: 1396.446 ms
```

**Comment by clarafu** [comment]
@zoetian and I looked into the web nodes to see if there was anything that stood out that could be causing high database cpu. We took a stack, heap and cpu profile for each of the two web nodes and wasn't able to find much. It looked like a portion of cpu and memory was spent on the listing all jobs, resources and teams. 

![Screenshot_2020-05-13 concourse inuse_space_heap](https://user-images.githubusercontent.com/19290703/81830583-3d5df200-950a-11ea-9dc0-f4af7be1e047.png)

![Screenshot_2020-05-13 concourse cpu](https://user-images.githubusercontent.com/19290703/81830567-3800a780-950a-11ea-88b6-8a846e6c57b0.png)

When we took a look at the stack dump, the first time we took the dump there were a lot of processes  (mostly from `SetCheckStepError` from lidar) that were waiting for a connection. This was probably because at the time that we took the stack dump, the backend connection pool limit was probably maxed and a lot of queries were unable to grab a connection. After we took a few more stack dumps, the connection pool was not maxed and everything seemed normal. The only ting that stood out was that there were a  few processes around build events that seemed to be around for quite a while (a couple of hours).

[stack.txt](https://github.com/concourse/concourse/files/4622889/stack.txt)
